### PR TITLE
Various small SQLAlchemy type improvements

### PR DESCRIPTION
**Pull request by srittau**

### CODE REVIEW

File: stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi [hunk 1]
```diff
@@ -1,21 +1,27 @@
...
 collections_abc = collections.abc
 
-EMPTY_SET: Any
+EMPTY_SET: frozenset[Any]
 
 class ImmutableContainer:
-    __delitem__: Any
-    __setitem__: Any
-    __setattr__: Any
+    def __delitem__(self, *arg: object, **kw: object) -> NoReturn: ...
```
Comment (JelleZijlstra) [inline stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi:18]: Would setting the arguments to NoReturn work better with typecheckers?

File: stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi [hunk 2]
```diff
@@ -27,31 +33,34 @@ class FacadeDict(ImmutableContainer, dict[Any, Any]):
...
+    def __dir__(self) -> list[str]: ...
+    def __add__(self, other: Iterable[_S]) -> list[_S | _T]: ...
+    def __setitem__(self, key: str, obj: _T) -> None: ...
+    def __getitem__(self, key: str) -> _T: ...
+    def __delitem__(self, key: str) -> None: ...
+    def __setattr__(self, key: str, obj: _T) -> None: ...
+    def __getattr__(self, key: str) -> _T: ...
+    def __contains__(self, key: str) -> bool: ...
+    def as_immutable(self) -> ImmutableProperties[_T]: ...
+    def update(self, value: Iterable[_T]) -> None: ...
```
Comment (JelleZijlstra) [inline stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi:49]: This calls dict.update, so it should take an `Iterable[tuple[str, _T]]` or a `SupportsKeysAndGetItem[str, _T]`.

File: stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi (hunk 1 above)
Comment (srittau) [reply stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi:18]: The arguments?

Comment (JelleZijlstra) [reply stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi:18]: Yes, calling a function with NoReturn argument will generally cause type checkers to show an error, which is what we'd want here.

Comment (srittau) [reply stubs/SQLAlchemy/sqlalchemy/util/_collections.pyi]: Considering the necessary `# type: ignore`s and the hacky nature of this, I'd prefer to keep using `object` (or `Any`).

**Pull request merged**